misc/clock.ru

```diff
@@ -3,19 +3,29 @@
 require "rack"
 require 'agoo'
 
+$ticking = true
+# Note that the sender can become invalid at any time if the connection is
+# closed. It will raise an exception on send if that occurs.
+$publish_thread = Thread.new {
+  while $ticking do
+    now = Time.now
+    Agoo.publish channel: :time, message: ("%02d:%02d:%02d" % [now.hour, now.min, now.sec])
+    sleep(1)
+  end
+}
```

Comment:
Good you put this here. I think we have a bit of a disconnect. Certainly for efficiency one thread updating a clock is better than many. I was trying to keep the example simple. What you are showing here though is the functionality I was trying to describe with `broadcast` as it hits all connections. Of course we would not want to specify Agoo or Iodine directly for a common API. Minor point at right now though. Concept is there.

I see the echo for `on_message` demonstrates the callback and the write though.

Reply:
Oh, I see.

I took the `publish` and `subscribe` method names from [wikipedia](https://en.wikipedia.org/wiki/Publish–subscribe_pattern) and [Redis](https://redis.io/commands#pubsub).

The ActionCable `broadcast` method name was probably chosen to avoid naming conflicts with the internal use of Redis for pub/sub... but the same name (broadcast) is also related to UDP and other network message broadcasting approaches, so I wasn't sure what you meant.

Anyway, since we're authoring the internal engine, I recommend that we keep to `publish` and `subscribe`, so Rails could use them internally without any naming conflicts.

What do you think?

Reply:
Yes, I agree that we don't want to specify Agoo or Iodine directly.

There should be some way to find the active Rack handler... shouldn't there.

We could monkey patch Rack to add:

```ruby
Rack.publish channel: :foo, message: "bar"
#or
Rack.subscribe(channel: :foo) {|msg| }
```

... but maybe there's a better way. What do you think?

Reply:
I am happy to play with different name combinations. I think we need two for writing. One to cover writing on a single connection and one to handle writes on potentially multiple connections.

Reply:
Conceptually, for publish, I agree. We need to figure out what the correct pub-sub model is though in regard to channel/topic/subject. 

For subscribe I'm not sure. It seems like you are blending the cross process messaging with the WebSockets.

Reply:
I agree.

Currently iodine uses `write` to write to a single object (reminiscent of `IO#write`) and `publish` to publish messages to a channel (which might have zero or more subscribers).

If you have other ideas, I'd be happy to listen, but if it's all the same to you than I would love to retain as much backwards compatibility as I can.

Reply:
> Conceptually, for publish, I agree. We need to figure out what the correct pub-sub model is though in regard to channel/topic/subject.

The names `channel` and `pattern` (in the current iodine implementation) come from Redis (I modeled my API to be as compatible with Redis as possible). We can probably use something else, but these names are already well known.

I know ActionCable uses the keyword `stream` to indicate the name (it maps directly to Redis `channels`)... but they do use the Class name `Channel` to indicate the subscription layer.

> For subscribe I'm not sure. It seems like you are blending the cross process messaging with the WebSockets.
 
I'm not sure I understand your reservation. Why is the subscription model limited to cross-process communication?

You can see [the iodine chat example here](https://github.com/boazsegev/iodine/blob/master/examples/redis.ru). IMHO, the API is quite easy to use (even if my code is cluttered).

Reply:
Ok, `write` and `publish` it is.

My concern with `channel` is that it implies a named connection between publisher and subscriber. I like the flexibility of a subject with wild cards like NATS or the first pub-sub implementation from TIBCO a couple of decades ago. Maybe what you describe as `pattern` is that. In any case if there is the notion of a subscribe call then I presume the client (browser) would send some kind of subscribe command to the handler on the Ruby side. That starts to get complicated. I prefer a simpler path as the pattern/filter for subscriptions. That also lets the Ruby code stay out of the way for direct browser subscription onto a messaging system like NATS or Redis.

Reply:
>  I presume the client (browser) would send some kind of subscribe command to the handler on the Ruby side.

Hmmm, not really.

This is definitely an option, but it's more common to have an application dictate which subscriptions a client is allowed or required to have.

For example, a client invites Alice and Bob to a group chat and the app will subscribe all three to a designated unique channel (probably `"group.chat.#{group.id}"` or another unique value).

>  I prefer a simpler path as the pattern/filter for subscriptions. 

I think the concept is valid, but it does hinder flexibility.

It also puts the choice of a channel in the client's hands, which might raise security concerns (what if the client's credentials are valid but the path was maliciously altered?).

This approach might also increase server resource consumption by requiring a single client to open a number of different connections (each with it's own subscriptions).

Last, but not least, I think this approach might conflict with the way apps are being designed today (where a single Websocket connection is shared for all subscriptions).

>  I like the flexibility of a subject with wild cards like NATS or the first pub-sub implementation from TIBCO a couple of decades ago. 

I don't know either of these technologies so well (sorry). My only role model was Redis and the technologies that were modeled to use it (ActionCable, socket.io, etc').

On the other hand, these are the technologies many developers are using these days, so using these terms gives us an advantage as far as adaptation is concerned ([Redis is the most loved database these days](https://insights.stackoverflow.com/survey/2018/#technology-most-loved-dreaded-and-wanted-databases), so adopting it's terminology will ease the learning curve).

With Redis, `pattern` allows for simple glob matching. `pattern` subscriptions are supported by Redis while `pattern` publishing isn't.

This is probably similar to `subject` in NAT, except that the `*` and `?` characters aren't reserved characters when subscribing to a `channel` (vs. a `pattern`), allowing the database to ignore the name and improve performance (no need to loop over the string and test if it's a pattern or not).

These also go to show what features are actually useful or in high demand. For example, the optimization signifies that most subscriptions aren't patterns (I only ever use patterns for tests or administrative tools).

> My concern with `channel` is that it implies a named connection between publisher and subscriber.

It's one way to think about it, except the `channel` might have any number of subscribers, including zero.

I think of it more like a TV channel than a connection.

Besides, as you well know, the parties really can't communicate with each other. Publishers have no knowledge of subscribers and subscribers have no idea who published the message (even though messages can claim to have a source). So I doubt there would be any confusion.